### ISSUE TITLE
Allow `FramedRead` user to choose the initial buffer size

### ISSUE THREAD

**Issue by stammw** [tokio-rs/tokio#2156]
## Subcrates

tokio-util 0.2.0

## Description

Allow `FramedRead` user to choose the initial buffer size writing something like: 

```rust
use tokio_util::codec:: FramedRead;

pub fn stream<T: AsyncRead>(stream: T) -> FramedRead<T, Self> {
    FramedRead::with_buffer(
        stream,
        MyDecoder::new(),
        BytesMut::with_capacity(65535)
    )
}
```

Sometimes, we can make a good guess of which size will be needed to be read in order to maximize the chance of getting a full frame at each `FrameDecoder::decode()` call. I see that `INITIAL_CAPACITY` for `FramedRead` is `8k`, which might be to small for some implementations.

Letting the user set this capacity could help him to reduce the chance of a reallocation when he knows frames end will be beyond those `8k`.

I hope I made my point clear enough, I'd be happy to submit a PR if you agree with it.

**Comment by LucioFranco** [comment]
@stammw I think this makes sense to provide a new fn to do this 👍

**Comment by stammw** [comment]
I'll submit a PR then.

**Comment by mzabaluev** [comment]
It looks like the immediate issue got fixed with #2215.

For a non-reallocating buffer that works well in the general case without manual tuning, I think a redesign would be needed to make use of something like https://github.com/mzabaluev/chunked-bytes for encoder output.
